woocommerce/woocommerce-paypal-payments
# Fix: Broken multi-level (nested) associative array values after getting submitted from checkout page

**Issue**: #306 

---

### Description
Instead of processing the submitted values using `explode` used a PHP function `parse_str` which can handle multi-level (nested) associative arrays.

### Steps to test:

1. Copy & paste the following block of code at the end of the file **functions.php** of child theme of your website
```php
/**
 * Add some HTML input fields on the checkout page
 *
 * @return void
 */
function ratnakar_custom_checkout_array_fields() {
	for ( $i = 0; $i < 3; $i++ ) {
		for ( $j = 0; $j < 2; $j++ ) {
			?>
			<input type="hidden" name="ratnakar_input[field_<?php echo esc_attr( $i ); ?>][]" value="ratnakar_input_value_<?php echo esc_attr( $i . '_' . $j ); ?>">
			<?php
		}
	}
}
add_action( 'woocommerce_checkout_after_customer_details', 'ratnakar_custom_checkout_array_fields' );

/**
 * Process the submitted values on the checkout page
 *
 * @return void
 */
function ratnakar_custom_checkout_processed_array_fields() {
	$logger  = wc_get_logger();
	$context = array( 'source' => 'ratnakar' );
	ob_start();
	var_dump( $_REQUEST );
	$logger->log( 'debug', 'ratnakar_input: ' . ob_get_clean() . ' ' . __FILE__ . ' ' . __LINE__, $context );
}
add_action( 'woocommerce_new_order', 'ratnakar_custom_checkout_processed_array_fields' );
```
2. Save the **functions.php** file
3. Now, go to the website, add a product to cart, navigate to the checkout page
4. On the checkout page, select **PayPal Payments** & click on the yellow button of the PayPal
5. Go to **WooCommerce > Status > Logs**
6. Look in the dropdown menu for log file name starting with '**ratnakar**'
7. Click on the **View** button to open the log
8. Go through the submitted form values, especially for the field `ratnakar_input`. You'll notice the square box that represents an array has become the part of the array key as string

### Result
After applying this fix, the submitted form values will be formatted correctly, as shown in the attached screenshot

![Screenshot from 2021-10-05 12-53-55](https://user-images.githubusercontent.com/3380686/135988564-9aa3b837-0666-42f4-81eb-e4501f61d567.png)

**Comparison in the results before & after applying the fix**

![Screenshot from 2021-10-05 14-05-58](https://user-images.githubusercontent.com/3380686/135989652-5f9b862b-4bb5-4a1d-a367-cf591508e486.png)

### Documentation
Don't think that this PR will require any change in the documentation

### Changelog entry
> Fix: Broken multi-level (nested) associative array values after getting submitted from checkout page

Closes #306
